### PR TITLE
Made Loss Multipliers as part of the parameter dictionary

### PR DESCRIPTION
**Pull request by jmcguigan10**

### CODE REVIEW

File: model_training/get_current_lr.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#################################################################
+# --------->>>| Function that always returns current learning
+# --------->>>| rate even for different schedulers
+# ###############################################################
+def current_lr(optim, sched=None):
+    # Varlid for modern schedulers: StepLR, CosineAnnealingLR, etc.
```
Comment (srichers) [inline model_training/get_current_lr.py]: spelling

File: model_training/get_current_lr.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    if sched is not None and hasattr(sched, "get_last_lr"):
+        try:
+            return sched.get_last_lr()[0]
+        except Exception:
+            pass
```
Comment (srichers) [inline model_training/get_current_lr.py:11]: If this fails the code should quit, not fail silently. I would prefer a single line instead of a try-except block.

File: model_training/get_current_lr.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    # always valid, including for ReduceLROnPlateau
+    return optim.param_groups[0]["lr"]
```
Comment (srichers) [inline model_training/get_current_lr.py:13]: If this is always valid, why not just use this one line for everything?

File: model_training/ml_trainmodel.py [hunk 4]
```diff
@@ -19,123 +20,156 @@
...
+
     # combine the datasets
     dataset_train = ConcatDataset(dataset_train_list)
-    dataset_test  = ConcatDataset(dataset_test_list )
-    
+    dataset_test = ConcatDataset(dataset_test_list)
+
     # create sampler and data loader for test data
-    sampler = WeightedRandomSampler(weights=weights, num_samples=parms["epoch_num_samples"], replacement=True)
+    sampler = WeightedRandomSampler(
```
Comment (srichers) [inline model_training/ml_trainmodel.py:40]: For this and previous line, try to keep whitespace changes as a separate PR - it significantly increases the speed at which a PR can be reviewed.

File: model_training/ml_trainmodel.py [hunk 5]
```diff
@@ -147,84 +181,162 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
+                fluxmag_true_test,
+                "test",
+                "fluxmag",
+                comparison_loss_fn,
+            )
+
             # train on making sure the model prediction is correct [direction]
-            train_loss = train_loss + contribute_loss(Fhat_pred_train, Fhat_true_train, "train", "direction", direction_loss_fn)
-            test_loss  = test_loss  + contribute_loss(Fhat_pred_test , Fhat_true_test , "test" , "direction", direction_loss_fn)
+            train_loss = train_loss + parms["direction_mult"] * contribute_loss(
```
Comment (srichers) [inline model_training/ml_trainmodel.py]: I think these parameters would be better named fully so that they are readable. It could also be good to have the first part of each be the same. For instance "loss_multiplier_direction" and "loss_multiplier_fluxmag", etc

**Review by srichers — CHANGES_REQUESTED** [review]
This is awesome, thanks. In the future, please try to minimize whitespace changes unless it's in a separate PR, since lots of whitespace changes makes it more difficult to parse through the PR to figure out what was actually changed. There are a few comments - could you address them and push your changes to this branch again? The PR will automatically update.

File: model_training/get_current_lr.py (hunk 2 above)
Comment (jmcguigan10) [reply model_training/get_current_lr.py:11]: What if I just added a print statement or something?

File: model_training/get_current_lr.py (hunk 3 above)
Comment (jmcguigan10) [reply model_training/get_current_lr.py:13]: I think it's just better practice to update the learning rate through the scheduler class and not the optimizer.

File: model_training/ml_trainmodel.py (hunk 5 above)
Comment (jmcguigan10) [reply model_training/ml_trainmodel.py]: I agree, I'll go ahead and change it.

File: model_training/get_current_lr.py (hunk 2 above)
Comment (srichers) [reply model_training/get_current_lr.py:11]: If we let things fail silently, there is a chance that we get results with hidden problems. I like failures to be loud unless we know we want to brush it under the rug.

File: model_training/get_current_lr.py (hunk 3 above)
Comment (srichers) [reply model_training/get_current_lr.py:13]: Sure, fair enough. We should force that to happen, then. I think there is a different standard of practice for application codes that are meant to be online and never go down and our type of calculations.
